File: gospider/core/curlconverter.go
package core

import (
	"bytes"
	"fmt"
	"os/exec"
	"sort"
	"strings"
)

// CurlToGoConverter 将 curl 命令转换为 Go 代码
type CurlToGoConverter struct {
	useCLI bool
}

type parsedCurlCommand struct {
	Method  string
	URL     string
	Headers map[string]string
	Data    string
}

// NewCurlToGoConverter 创建转换器
func NewCurlToGoConverter() *CurlToGoConverter {
	return &CurlToGoConverter{
		useCLI: true, // Go 版本主要使用 CLI 工具
	}
}

// Convert 将 curl 命令转换为 Go 代码
func (c *CurlToGoConverter) Convert(curlCommand string) (string, error) {
	// 使用 curlconverter 进行转换
	var cmd *exec.Cmd
	
	// 尝试直接使用 curlconverter（如果已全局安装）
	cmd = exec.Command("curlconverter", "--language", "go", "-")
	cmd.Stdin = strings.NewReader(curlCommand)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		// 如果失败，尝试使用 npx
		cmd = exec.Command("npx", "curlconverter", "--language", "go", "-")
		cmd.Stdin = strings.NewReader(curlCommand)
		cmd.Stdout = &stdout
		cmd.Stderr = &stderr
		stdout.Reset()
		stderr.Reset()
		
		err = cmd.Run()
		if err != nil {
			return "", fmt.Errorf("curlconverter 错误: %v, stderr: %s", err, stderr.String())
		}
	}

	return stdout.String(), nil
}

// ConvertToHTTP 转换为使用 net/http 的代码
func (c *CurlToGoConverter) ConvertToHTTP(curlCommand string) (string, error) {
	goCode, err := c.Convert(curlCommand)
	if err != nil {
		return "", err
	}
	
	// 添加必要的 import
	if !strings.Contains(goCode, "package") {
		goCode = "package main\n\n" + goCode
	}
	
	return goCode, nil
}

// ConvertToResty 转换为使用 go-resty 的代码（更简洁的 HTTP 客户端）
func (c *CurlToGoConverter) ConvertToResty(curlCommand string) string {
	parsed, err := parseCurlCommand(curlCommand)
	if err != nil {
		return fmt.Sprintf(`package main

import (
	"fmt"
	"github.com/go-resty/resty/v2"
)

func main() {
	client := resty.New()

	// 无法可靠解析 curl 命令，保留原始输入供人工调整。
	// curl: %s
	resp, err := client.R().
		Get("https://httpbin.org/get")

	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	fmt.Println("Response:", resp.Status())
	fmt.Println("Body:", string(resp.Body()))
}
`, curlCommand)
	}

	var builder strings.Builder
	builder.WriteString("package main\n\n")
	builder.WriteString("import (\n")
	builder.WriteString("\t\"fmt\"\n")
	builder.WriteString("\t\"github.com/go-resty/resty/v2\"\n")
	builder.WriteString(")\n\n")
	builder.WriteString("func main() {\n")
	builder.WriteString("\tclient := resty.New()\n")
	builder.WriteString("\treq := client.R()\n")

	if len(parsed.Headers) > 0 {
		keys := make([]string, 0, len(parsed.Headers))
		for key := range parsed.Headers {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		builder.WriteString("\treq = req.SetHeaders(map[string]string{\n")
		for _, key := range keys {
			builder.WriteString(fmt.Sprintf("\t\t%q: %q,\n", key, parsed.Headers[key]))
		}
		builder.WriteString("\t})\n")
	}

	if parsed.Data != "" {
		builder.WriteString(fmt.Sprintf("\treq = req.SetBody(%q)\n", parsed.Data))
	}

	builder.WriteString(fmt.Sprintf("\tresp, err := req.Execute(%q, %q)\n", parsed.Method, parsed.URL))
	builder.WriteString("\tif err != nil {\n")
	builder.WriteString("\t\tfmt.Println(\"Error:\", err)\n")
	builder.WriteString("\t\treturn\n")
	builder.WriteString("\t}\n\n")
	builder.WriteString("\tfmt.Println(\"Response:\", resp.Status())\n")
	builder.WriteString("\tfmt.Println(\"Body:\", resp.String())\n")
	builder.WriteString("}\n")
	return builder.String()
}

// InstallCurlconverter 安装 curlconverter CLI 工具
func InstallCurlconverter() error {
	// 检查 npm 是否可用
	_, err := exec.LookPath("npm")
	if err != nil {
		return fmt.Errorf("npm 未安装: %v", err)
	}
	
	// 安装 curlconverter
	cmd := exec.Command("npm", "install", "-g", "curlconverter")
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("安装失败: %v, output: %s", err, string(output))
	}
	
	return nil
}

// CurlToGo 便捷函数
func CurlToGo(curlCommand string) (string, error) {
	converter := NewCurlToGoConverter()
	return converter.Convert(curlCommand)
}

func parseCurlCommand(curlCommand string) (*parsedCurlCommand, error) {
	tokens, err := tokenizeCurlCommand(curlCommand)
	if err != nil {
		return nil, err
	}
	if len(tokens) == 0 {
		return nil, fmt.Errorf("empty curl command")
	}
	if strings.EqualFold(tokens[0], "curl") {
		tokens = tokens[1:]
	}

	parsed := &parsedCurlCommand{
		Method:  "GET",
		Headers: make(map[string]string),
	}
	var dataParts []string

	for i := 0; i < len(tokens); i++ {
		token := tokens[i]
		switch token {
		case "-X", "--request":
			if i+1 >= len(tokens) {
				return nil, fmt.Errorf("missing request method")
			}
			i++
			parsed.Method = strings.ToUpper(tokens[i])
		case "-H", "--header":
			if i+1 >= len(tokens) {
				return nil, fmt.Errorf("missing header value")
			}
			i++
			parts := strings.SplitN(tokens[i], ":", 2)
			if len(parts) == 2 {
				parsed.Headers[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
			}
		case "-d", "--data", "--data-raw", "--data-binary", "--data-urlencode":
			if i+1 >= len(tokens) {
				return nil, fmt.Errorf("missing data value")
			}
			i++
			dataParts = append(dataParts, tokens[i])
			if parsed.Method == "GET" {
				parsed.Method = "POST"
			}
		default:
			if strings.HasPrefix(token, "http://") || strings.HasPrefix(token, "https://") {
				parsed.URL = token
			}
		}
	}

	if parsed.URL == "" {
		return nil, fmt.Errorf("missing target url")
	}
	if len(dataParts) > 0 {
		parsed.Data = strings.Join(dataParts, "&")
	}
	return parsed, nil
}

func tokenizeCurlCommand(input string) ([]string, error) {
	var tokens []string
	var current strings.Builder
	var quote rune
	escaped := false

	flush := func() {
		if current.Len() > 0 {
			tokens = append(tokens, current.String())
			current.Reset()
		}
	}

	for _, ch := range input {
		switch {
		case escaped:
			current.WriteRune(ch)
			escaped = false
		case ch == '\\':
			escaped = true
		case quote != 0:
			if ch == '\\' {
				escaped = true
			} else if ch == quote {
				quote = 0
			} else {
				current.WriteRune(ch)
			}
		case ch == '\'' || ch == '"':
			quote = ch
		case ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r':
			flush()
		default:
			current.WriteRune(ch)
		}
	}

	if escaped {
		current.WriteRune('\\')
	}
	if quote != 0 {
		return nil, fmt.Errorf("unterminated quote in curl command")
	}
	flush()
	return tokens, nil
}

File: rustspider/src/curlconverter.rs
//! curlconverter 集成模块
//! 将 curl 命令转换为 Rust 代码

use std::collections::BTreeMap;
use std::io::Write;
use std::process::Command;

/// curl 到 Rust 转换器
pub struct CurlToRustConverter {
    use_cli: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct ParsedCurlCommand {
    method: String,
    url: String,
    headers: BTreeMap<String, String>,
    data: Option<String>,
}

impl CurlToRustConverter {
    /// 创建新的转换器实例
    pub fn new() -> Self {
        Self {
            use_cli: true, // Rust 版本使用 CLI 工具
        }
    }

    /// 将 curl 命令转换为 Rust 代码
    pub fn convert(&self, curl_command: &str) -> Result<String, String> {
        // 尝试直接使用 curlconverter
        let result = self.run_curlconverter("curlconverter", curl_command);

        // 如果失败，尝试使用 npx
        if result.is_err() {
            return self.run_curlconverter("npx", curl_command);
        }

        result
    }

    /// 运行 curlconverter 命令
    fn run_curlconverter(&self, command: &str, curl_command: &str) -> Result<String, String> {
        let mut cmd = Command::new(command);

        if command == "npx" {
            cmd.arg("curlconverter");
        }

        cmd.arg("--language")
            .arg("rust")
            .arg("-") // 从 stdin 读取
            .stdin(std::process::Stdio::piped())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped());

        let mut child = cmd
            .spawn()
            .map_err(|e| format!("启动 {} 失败: {}", command, e))?;

        // 写入 curl 命令到 stdin
        if let Some(ref mut stdin) = child.stdin {
            stdin
                .write_all(curl_command.as_bytes())
                .map_err(|e| format!("写入 stdin 失败: {}", e))?;
        }

        let output = child
            .wait_with_output()
            .map_err(|e| format!("等待子进程失败: {}", e))?;

        if output.status.success() {
            let rust_code = String::from_utf8_lossy(&output.stdout);
            Ok(rust_code.to_string())
        } else {
            let error = String::from_utf8_lossy(&output.stderr);
            Err(format!("{} 错误: {}", command, error))
        }
    }

    /// 转换为使用 reqwest 的代码
    pub fn convert_to_reqwest(&self, curl_command: &str) -> Result<String, String> {
        let rust_code = self.convert(curl_command)?;

        // 确保包含必要的依赖
        let mut result = String::new();

        if !rust_code.contains("[dependencies]") {
            result.push_str("// Cargo.toml 依赖:\n");
            result.push_str("// [dependencies]\n");
            result.push_str("// reqwest = { version = \"0.11\", features = [\"json\"] }\n");
            result.push_str("// tokio = { version = \"1\", features = [\"full\"] }\n\n");
        }

        result.push_str(&rust_code);
        Ok(result)
    }

    /// 转换为使用 ureq 的代码（同步 HTTP 客户端）
    pub fn convert_to_ureq(&self, curl_command: &str) -> String {
        let Ok(parsed) = parse_curl_command(curl_command) else {
            return format!(
                r#"// 使用 ureq 的同步 HTTP 客户端
// Cargo.toml: ureq = "2.9"

use std::io::Read;

fn main() -> Result<(), Box<dyn std::error::Error>> {{
    // curl: {}

    let mut resp = ureq::get("https://httpbin.org/get")
        .set("Accept", "application/json")
        .call()?;

    let mut body = String::new();
    resp.into_reader().read_to_string(&mut body)?;

    println!("Response: {{}}", body);
    Ok(())
}}
"#,
                curl_command
            );
        };

        let (constructor, uses_generic_request) = ureq_constructor(&parsed.method);
        let headers = if parsed.headers.is_empty() {
            String::new()
        } else {
            parsed
                .headers
                .iter()
                .map(|(key, value)| format!("\n        .set({key:?}, {value:?})"))
                .collect::<String>()
        };

        let request_builder = if uses_generic_request {
            format!("ureq::request({:?}, {:?})", parsed.method, parsed.url)
        } else {
            format!("{constructor}({:?})", parsed.url)
        };

        let send_code = if let Some(data) = parsed.data.as_ref() {
            format!("let mut resp = {request_builder}{headers}\n        .send_string({data:?})?;")
        } else {
            format!("let mut resp = {request_builder}{headers}\n        .call()?;")
        };

        format!(
            r#"// 使用 ureq 的同步 HTTP 客户端
// Cargo.toml: ureq = "2.9"

use std::io::Read;

fn main() -> Result<(), Box<dyn std::error::Error>> {{
    // curl: {}

    {}

    let mut body = String::new();
    resp.into_reader().read_to_string(&mut body)?;

    println!("Response: {{}}", body);
    Ok(())
}}
"#,
            curl_command, send_code
        )
    }

    /// 安装 curlconverter CLI
    pub fn install_curlconverter() -> Result<(), String> {
        let output = Command::new("npm")
            .arg("install")
            .arg("-g")
            .arg("curlconverter")
            .output()
            .map_err(|e| format!("安装失败: {}", e))?;

        if output.status.success() {
            Ok(())
        } else {
            let error = String::from_utf8_lossy(&output.stderr);
            Err(format!("安装失败: {}", error))
        }
    }
}

impl Default for CurlToRustConverter {
    fn default() -> Self {
        Self::new()
    }
}

/// 便捷函数：将 curl 命令转换为 Rust 代码
pub fn curl_to_rust(curl_command: &str) -> Result<String, String> {
    let converter = CurlToRustConverter::new();
    converter.convert(curl_command)
}

fn parse_curl_command(curl_command: &str) -> Result<ParsedCurlCommand, String> {
    let mut tokens = tokenize_curl_command(curl_command)?;
    if tokens.is_empty() {
        return Err("empty curl command".to_string());
    }
    if tokens
        .first()
        .map(|token| token.eq_ignore_ascii_case("curl"))
        == Some(true)
    {
        tokens.remove(0);
    }

    let mut parsed = ParsedCurlCommand {
        method: "GET".to_string(),
        url: String::new(),
        headers: BTreeMap::new(),
        data: None,
    };
    let mut data_parts = Vec::new();
    let mut index = 0usize;
    while index < tokens.len() {
        match tokens[index].as_str() {
            "-X" | "--request" => {
                index += 1;
                if index >= tokens.len() {
                    return Err("missing request method".to_string());
                }
                parsed.method = tokens[index].to_uppercase();
            }
            "-H" | "--header" => {
                index += 1;
                if index >= tokens.len() {
                    return Err("missing header value".to_string());
                }
                if let Some((key, value)) = tokens[index].split_once(':') {
                    parsed
                        .headers
                        .insert(key.trim().to_string(), value.trim().to_string());
                }
            }
            "-d" | "--data" | "--data-raw" | "--data-binary" | "--data-urlencode" => {
                index += 1;
                if index >= tokens.len() {
                    return Err("missing data value".to_string());
                }
                data_parts.push(tokens[index].clone());
                if parsed.method == "GET" {
                    parsed.method = "POST".to_string();
                }
            }
            token if token.starts_with("http://") || token.starts_with("https://") => {
                parsed.url = token.to_string();
            }
            _ => {}
        }
        index += 1;
    }

    if parsed.url.is_empty() {
        return Err("missing target url".to_string());
    }
    if !data_parts.is_empty() {
        parsed.data = Some(data_parts.join("&"));
    }

    Ok(parsed)
}

fn tokenize_curl_command(input: &str) -> Result<Vec<String>, String> {
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut quote: Option<char> = None;
    let mut escaped = false;

    let flush = |tokens: &mut Vec<String>, current: &mut String| {
        if !current.is_empty() {
            tokens.push(current.clone());
            current.clear();
        }
    };

    for ch in input.chars() {
        if escaped {
            current.push(ch);
            escaped = false;
            continue;
        }

        match quote {
            Some(active_quote) if ch == active_quote => {
                quote = None;
            }
            Some(_) if ch == '\\' => {
                escaped = true;
            }
            Some(_) => current.push(ch),
            None if ch == '\\' => escaped = true,
            None if ch == '"' || ch == '\'' => quote = Some(ch),
            None if ch.is_whitespace() => flush(&mut tokens, &mut current),
            None => current.push(ch),
        }
    }

    if escaped {
        current.push('\\');
    }
    if quote.is_some() {
        return Err("unterminated quote in curl command".to_string());
    }
    flush(&mut tokens, &mut current);
    Ok(tokens)
}

fn ureq_constructor(method: &str) -> (&'static str, bool) {
    match method {
        "GET" => ("ureq::get", false),
        "POST" => ("ureq::post", false),
        "PUT" => ("ureq::put", false),
        "PATCH" => ("ureq::patch", false),
        "DELETE" => ("ureq::delete", false),
        _ => ("ureq::request", true),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_convert_simple_curl() {
        let curl_cmd = r#"curl -X GET "https://httpbin.org/get""#;
        let converter = CurlToRustConverter::new();
        // 注意：这需要 npm 和 curlconverter 已安装
        let result = converter.convert(curl_cmd);
        // 测试可能失败，如果 curlconverter 未安装
        if let Ok(output) = result {
            assert!(output.contains("reqwest"));
        }
    }

    #[test]
    fn test_parse_curl_command_supports_headers_and_body() {
        let parsed = parse_curl_command(
            r#"curl -X POST "https://example.com/api" -H "Accept: application/json" -d "{\"name\":\"ultra\"}""#,
        )
        .expect("parsed curl command");

        assert_eq!(parsed.method, "POST");
        assert_eq!(parsed.url, "https://example.com/api");
        assert_eq!(
            parsed.headers.get("Accept").map(String::as_str),
            Some("application/json")
        );
        assert_eq!(parsed.data.as_deref(), Some(r#"{"name":"ultra"}"#));
    }

    #[test]
    fn test_convert_to_ureq_uses_parsed_target() {
        let converter = CurlToRustConverter::new();
        let code = converter.convert_to_ureq(
            r#"curl -X POST "https://example.com/api" -H "Content-Type: application/json" -d "{\"name\":\"ultra\"}""#,
        );

        assert!(code.contains("https://example.com/api"));
        assert!(code.contains("ureq::post"));
        assert!(code.contains(".set(\"Content-Type\", \"application/json\")"));
        assert!(code.contains(".send_string(\"{\\\"name\\\":\\\"ultra\\\"}\")"));
    }
}
